qrcode_detection_package/main.py

```diff
@@ -1,56 +1,358 @@
+                img_path = f'images/{timestamp}.jpg'
+                cv2.imwrite(img_path, captured_image)
+                self.get_logger().info(
+                    f"Saved image of detected QR-Code as: {img_path}")
+
+                # create QRCodeInfo message to publish on qr_codes topic
+                msg = QRCodeInfo()
+                msg.time_stamp = self.get_clock().now().to_msg()
+                msg.sender_id = "qr_code_scanner_node"
+                msg.qr_code_content = qr_code_content
+                msg.qrcode_position_x = qrcode_center_x
+                msg.qrcode_position_y = qrcode_center_y
+
+                self.qr_publisher.publish(msg)
+                self.get_logger().info("Published QR code info")
+
+                # create dict for sending job finished message
+                payload = {"marker": str(qr_code_content)}
+                self._job_finished_custom_(CommonNode.EXIT_SUCCESS, payload)
+                self.get_logger().info("Published job finished message")
+
+            else:
+                self.get_logger().info("No QR Code found")
+        else:
+            self.get_logger().info("Could not capture image")
 
 
 def main(args=None):
+    """
+    The main function initializes the node and runs the state machine over the lifetime of the node.
+
+    This function initializes the ROS 2 node, creates an instance of the QRCodeScannerNode class,
+    and starts the image processing loop. Then the state machine decides what the node does depending on the internal state.
+    The states are:
+    - "ready": The node is running and waits for a control message which activates it. There happens no image
+               capturing or qr code scanning in this state
+    - "searching": in this state the node continuously takes images and uses the OpenCV library to scan them for
+                   QR-Codes. If a valid code is found, the nodes publishes its contents and position and switches
+                   back to the state "ready"
+    It handles the cleanup operations before shutting down the node.
+
+    Args: 
+        args: Command-line arguments. Default is None.
+
+    Returns: 
+        None
+    """
     rclpy.init(args=args)
     node_id = 'qr_code_scanner_node'
     qr_code_scanner_node = QRCodeScannerNode(node_id)
 
-    try:
-        qr_code_scanner_node.process_images()
-    finally:
-        qr_code_scanner_node.destroy_node()
-        rclpy.shutdown()
+    # start excecuting state machine until node gets destroyed
+    while True:
+        # check if node is in state "ready"
+        # in this state the node waits for the control message to activate the node
+        if (qr_code_scanner_node.node_state == "ready"):
+            qr_code_scanner_node.get_logger().info("Node is in state ready")
+            # if the node got activated it sets its internal state to searching
+            if (qr_code_scanner_node.active):
+                qr_code_scanner_node.set_state("searching")
```

Comment:
You could also think about moving the code for each state into its own function as to increase legibility, but that is up to you.

Reply:
I see the point but i think in this case the code for each state is quite limited already so that looking up the function might just be about same level of legibility so it wouldn`t be a massive change imo